test/cbmc/sources/core_sntp_cbmc_state.c

```diff
@@ -0,0 +1,115 @@
+/*
+ * coreSNTP v1.0.0
+ * Copyright (C) 2021 Amazon.com, Inc. or its affiliates.  All Rights Reserved.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy of
+ * this software and associated documentation files (the "Software"), to deal in
+ * the Software without restriction, including without limitation the rights to
+ * use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+ * the Software, and to permit persons to whom the Software is furnished to do so,
+ * subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in all
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+ * FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
+ * COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
+ * IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+ * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+ */
+
+/**
+ * @file core_sntp_cbmc_state.c
+ * @brief Implements the functions defined in core_sntp_cbmc_state.h.
+ */
+#include <stdint.h>
+#include <stdlib.h>
+#include "core_sntp_client.h"
+#include "core_sntp_cbmc_state.h"
+#include "core_sntp_stubs.h"
+
+SntpContext_t * allocateCoreSntpContext( SntpContext_t * pContext )
+{
+    SntpServerInfo_t * pTimeServers;
+    size_t currentServerIndex;
+    size_t numOfServers;
+    uint32_t serverResponseTimeoutMs;
+    uint8_t * pNetworkBuffer;
+    size_t bufferSize;
+    SntpResolveDns_t resolveDnsFunc;
+    SntpGetTime_t getTimeFunc;
+    SntpSetTime_t setTimeFunc;
+    UdpTransportInterface_t * pNetworkIntf;
+    SntpAuthenticationInterface_t * pAuthIntf;
+    SntpStatus_t sntpStatus = SntpSuccess;
+
+    if( pContext == NULL )
+    {
+        pContext = malloc( sizeof( SntpContext_t ) );
+    }
+
+    __CPROVER_assume( numOfServers < CBMC_MAX_OBJECT_SIZE );
+    __CPROVER_assume( serverResponseTimeoutMs < CBMC_MAX_OBJECT_SIZE );
+    __CPROVER_assume( currentServerIndex < CBMC_MAX_OBJECT_SIZE );
+
+    if( numOfServers == 0 )
+    {
+        pTimeServers = NULL;
+    }
+    else
+    {
+        pTimeServers = malloc( numOfServers * sizeof( SntpServerInfo_t ) );
+    }
+
+    if( pTimeServers != NULL )
+    {
+        __CPROVER_assume( pTimeServers->serverNameLen < CBMC_MAX_OBJECT_SIZE );
+        __CPROVER_assume( pTimeServers->port < CBMC_MAX_OBJECT_SIZE );
+        pTimeServers->pServerName = malloc( pTimeServers->serverNameLen );
+    }
```

Comment:
This just contains assumptions for the first object in the allocated list of `SntpServerInfo_t` servers. Should this be a loop instead to enforce these assumptions on all the allocated elements of the list?
```suggestion
    if( pTimeServers != NULL )
    {
       for( size_t i = 0; i < numOfServers: i++ )
       {
           __CPROVER_assume( pTimeServers[i].serverNameLen < CBMC_MAX_OBJECT_SIZE );
           __CPROVER_assume( pTimeServers[i].port < CBMC_MAX_OBJECT_SIZE );
           pTimeServers[i].pServerName = malloc( pTimeServers[i].serverNameLen );
        }
    }
```